### DISCUSSION TITLE
InputTextMultine crash (when using ImGuiTabBarFlags_AutoSelectNewTabs and ImGuiTabItemFlags_SetSelected)

### DISCUSSION THREAD

**Discussion by WarEagle451** [ocornut/imgui#4522] [New Users Build/Link/Run issues ONLY!]
Version: 1.84 WIP
Branch: Docking

Back-ends: imgui_impl_opengl3.cpp + imgui_impl_glfw.cpp
Operating System: Windows 10, 19042

Issue:
When a tab bar has ImGuiTabBarFlags_AutoSelectNewTabs and a new tab is added with the ImGuiTabItemFlags_SetSelected flag it causes a crash

To reproduce:
```
ImGui::Begin("Window");
ImGui::BeginTabBar("##Tabs", ImGuiTabBarFlags_AutoSelectNewTabs);
if (ImGui::BeginTabItem("Tab", nullptr, ImGuiTabItemFlags_SetSelected))
{
	ImGui::EndTabItem();
}
ImGui::EndTabBar();
ImGui::End();
```

**Comment by PathogenDavid** [comment]
Can you share a stack trace of the crash? Your code does not crash for me.

**Comment by WarEagle451** [comment]
I haven't done this before so I hope this is correct.

>	Editor.exe!std::string::capacity() Line 3963	C++
 	Editor.exe!Mythos::Editor::Gui::InputTextMultiline(const std::string & label, std::string * str, const ImVec2 & size, int flags, int(*)(ImGuiInputTextCallbackData *) callback, void * user_data) Line 178	C++
 	Editor.exe!Mythos::Editor::Panels::TextEditorPanel::OnRender() Line 94	C++
 	Editor.exe!Mythos::Editor::PanelManager::OnRender() Line 71	C++
 	Editor.exe!Mythos::Editor::EditorLayer::OnImGuiRender() Line 227	C++
 	Editor.exe!Mythos::Application::Run() Line 97	C++
 	Editor.exe!main(int argc, char * * argv) Line 23	C++
 	Editor.exe!invoke_main() Line 79	C++
 	Editor.exe!__scrt_common_main_seh() Line 288	C++
 	Editor.exe!__scrt_common_main() Line 331	C++
 	Editor.exe!mainCRTStartup(void * __formal) Line 17	C++
 	kernel32.dll!00007ffcaa387034()	Unknown
 	ntdll.dll!00007ffcabb42651()	Unknown

Also so I just realized something weird, that code doesn't cause a crash if I put it in other places in my project (sorry about that) but only in the spot I'm trying to use it. Additionally I can still reproduce it by having both flags set, with either or I don't get a crash. I can share a non simplified version of my code for more context if you would like

**Comment by ocornut** [comment]
OK so this has nothing to do with Tabs.
Please investigate in more details because it could be simply that you are passing e.g. an incorrect std::string pointer to `InputTextMultiline()` in that situation. I'll move this to Discussion as it likely not a Dear ImGui issue.

**Comment by WarEagle451** [comment]
I dug more and you're correct but this confuses me even further because it should be set for the input text.
I see now on the first pass that the selected file is not set until after the tab bar is created but what makes no sense is why removing the ImGuiTabBarFlags_AutoSelectNewTabs flag somehow allows the pointer to the selected file to be invalid somehow?

Here is a more detailed version of what I am doing, I tested this and it does recreate the crash
```
	bool InputTextMultiline(const std::string& label, std::string* str, const ImVec2& size, ImGuiInputTextFlags flags, ImGuiInputTextCallback callback, void* user_data)
	{
		IM_ASSERT((flags & ImGuiInputTextFlags_CallbackResize) == 0);
		flags |= ImGuiInputTextFlags_CallbackResize;

		InputTextCallback_UserData cb_user_data{};
		cb_user_data.Str = str;
		cb_user_data.ChainCallback = callback;
		cb_user_data.ChainCallbackUserData = user_data;
		return ImGui::InputTextMultiline(label.c_str(), (char*)str->c_str(), str->capacity() + 1, size, flags, InputTextCallback, &cb_user_data);
	}

	struct File
	{
		File(const std::string& name, const std::string& data, ImGuiTabItemFlags flags) :
			Name(name),
			Data(data),
			Flags(flags) {}
		std::string Name, Data;
		ImGuiTabItemFlags Flags;
	};

	static std::vector<std::shared_ptr<File>> g_Files;
	static std::shared_ptr<File> g_Selected;

	void Render()
	{
		if (ImGui::Begin("Window"))
		{
			if (!g_Selected)
				g_Files.push_back(std::make_shared<File>("A File", "Words", ImGuiTabItemFlags_SetSelected));
			
			ImGui::BeginTabBar("TabBar", ImGuiTabBarFlags_AutoSelectNewTabs); // <- Why does removing (ImGuiTabBarFlags_AutoSelectNewTabs) fix the crash it shouldn't?
			for (auto& f : g_Files)
			{
				if (ImGui::BeginTabItem(f->Name.data(), nullptr, f->Flags))
				{
					if (f != g_Selected) // Because if the tab is active that should be the selected file
						g_Selected = f;
					if (f->Flags & ImGuiTabItemFlags_SetSelected)
						f->Flags &= ~ImGuiTabItemFlags_SetSelected;
					ImGui::EndTabItem();
				}
			}
			ImGui::EndTabBar();
			
			InputTextMultiline("text", &g_Selected->Data); // Same as ImGui's string version of InputText
		}
		ImGui::End();
	}
```
